### PR TITLE
Improve error msg when a field name contains only white spaces

### LINKED ISSUE #27701: Improve error message for whitespace-only JSON field names
**Elasticsearch version** : 5.2.1

**Plugins installed**:  none

**JVM version**:
openjdk version "1.8.0_152"
OpenJDK Runtime Environment (Zulu 8.25.0.1-macosx) (build 1.8.0_152-b16)
OpenJDK 64-Bit Server VM (Zulu 8.25.0.1-macosx) (build 25.152-b16, mixed mode)

**OS version**:
Darwin Jorg-Prantes-MacBook-Pro.local 16.7.0 Darwin Kernel Version 16.7.0: Wed Oct  4 00:17:00 PDT 2017; root:xnu-3789.71.6~1/RELEASE_X86_64 x86_64

**Description of the problem including expected versus actual behavior**:
The JSON format with a space as field name that was accepted up to Elastiscearch 5.2.0 should also be accepted by Elasticsearch 5.2.1+

This could be related to https://github.com/elastic/elasticsearch/pull/22891

The fix seems to forbid space as field name, but the intention was only to forbid empty field names.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1.  Putting this JSON to Elasticsearch with curl 
```
curl -XPUT 'http://localhost:9200/testindex/1/1' -d '
{
  "600": [
    {
      "07": [
        {
          "a": "Supports."
        }
      ]
    },
    {
      "07": [
        {
          "a": "."
        },
        {
          " ": "Surfaces <Künstlergemeinschaft>"
        }
      ]
    }
  ]
}
'
```
 2. Examine server log.
 3. There should be no error.

**Provide logs (if relevant)**:

```
java.lang.IllegalArgumentException: object field starting or ending with a [.] makes object resolution ambiguous: [ ]
	at org.elasticsearch.index.mapper.DocumentParser.splitAndValidatePath(DocumentParser.java:182) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.getMapper(DocumentParser.java:922) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseValue(DocumentParser.java:573) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:395) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:372) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:446) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:466) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseNonDynamicArray(DocumentParser.java:554) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseArray(DocumentParser.java:510) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:384) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:372) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:446) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:466) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseNonDynamicArray(DocumentParser.java:554) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseArray(DocumentParser.java:510) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:384) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:372) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:93) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:66) ~[elasticsearch-5.2.1.jar:5.2.1]
```

### PR DESCRIPTION
**Pull request by olcbean**
Till now if a field name is empty or contains only white spaces, the error msg is `object field starting or ending with a [.] makes object resolution ambiguous`.

This PR explicitly check if the field name contains only white spaces and result in the error `object field cannot contain only white spaces`.

Fixes #27701

CC @danielmitterdorfer @dakrone

### CODE REVIEW

**Comment by cbuescher** [comment]
@elasticmachine test this please

File: core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 1]
```diff
@@ -180,6 +180,11 @@ private static MapperParsingException wrapInMapperParsingException(SourceToParse
             String[] parts = fullFieldPath.split("\\.");
             for (String part : parts) {
                 if (Strings.hasText(part) == false) {
+                    // check if the field name contains only white spaces
+                    if (Strings.isEmpty(part) == false) {
+                        throw new IllegalArgumentException(
+                                "object field cannot contain only white spaces: ['" + fullFieldPath + "']");
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java]: this should be "white spaces" -> "whitespace"

File: core/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java [hunk 2]
```diff
@@ -1377,6 +1377,23 @@ public void testDynamicFieldsStartingAndEndingWithDot() throws Exception {
...
+                .field("a", 1).field(" ", 2)
+                .endObject()
+                .endObject().endArray()
+                .endObject().bytes();
+
+        IllegalArgumentException emptyFieldNameException = expectThrows(IllegalArgumentException.class,
+                () -> client().prepareIndex("idx", "type").setSource(bytes, XContentType.JSON).get());
+
+        assertThat(emptyFieldNameException.getMessage(), containsString(
+                "object field cannot contain only white spaces: ['top.aoeu. ']"));
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java]: "white spaces" -> "whitespace"

**Review by dakrone — CHANGES_REQUESTED** [review]
I left two comments about the messages

**Comment by olcbean** [comment]
@dakrone thanks! I just modified the msg.

**Comment by dakrone** [comment]
Merged, thanks @olcbean!

**Pull request merged**